src/lib/update-lock.ts

```diff
@@ -96,69 +87,93 @@ function dispose(
-													return lockfile.unlock(tmpLockName);
-												}
-											})
-												.then(() => {
-													return lockfile.lock(tmpLockName, LOCKFILE_UID);
-												})
-												// If lockfile exists, throw a user-friendly error.
-												// Otherwise throw the error as-is.
-												// This will interrupt the call to Bluebird.using, so
-												// dispose needs to be called even though it's referenced
-												// by .disposer later.
-												.catch((error) => {
-													return dispose(release, appId).throw(
-														lockfile.LockfileExistsError.is(error)
-															? new UpdatesLockedError(
-																	`Lockfile exists for ${JSON.stringify({
-																		serviceName,
-																		appId,
-																	})}`,
-															  )
-															: (error as Error),
-													);
-												})
-										);
-									},
-								);
-							});
-					})
-					.disposer((release: () => void) => dispose(release, appId));
-			})
-			.catch((err) => {
-				throw new InternalInconsistencyError(
-					`Error getting lockOverride config value: ${err?.message ?? err}`,
-				);
-			});
-	};
+	// Sort appIds so they are always locked in the same sequence
+	const sortedIds = appIdsToLock.sort();
 
-	const disposer = takeTheLock();
+	let lockOverride: boolean;
+	try {
+		lockOverride = await config.get('lockOverride');
+	} catch (err) {
+		throw new InternalInconsistencyError(
+			`Error getting lockOverride config value: ${err?.message ?? err}`,
+		);
+	}
 
-	return Bluebird.using(disposer, fn as () => PromiseLike<T>);
+	const releases = new Map<number, () => void>();
+	try {
+		for await (const id of sortedIds) {
+			const lockDir = getPathOnHost(lockPath(id));
+			// Acquire write lock for appId
+			releases.set(id, await writeLock(id));
+			// Get list of service folders in lock directory
+			let serviceFolders: string[] = [];
+			try {
+				serviceFolders = await fs.readdir(lockDir);
+			} catch (err) {
+				if (ENOENT(err)) {
+					// Default to empty list of folders
+					serviceFolders = [];
+				} else {
+					throw err;
+				}
+			}
+			// Attempt to create a lock for each service
+			for await (const service of serviceFolders) {
+				await lockService(id, service, { force, lockOverride });
+			}
+		}
+		// Resolve the function passed
+		return Promise.resolve(fn());
+	} finally {
+		// Cleanup locks
+		for await (const [id, release] of releases.entries()) {
+			// Release lock on application
+			release();
+			// Dispose of locks in each service within the application
+			await dispose(id);
+		}
+	}
+}
+
+type LockOptions = {
+	force?: boolean;
+	lockOverride?: boolean;
+};
+
+async function lockService(
+	appId: number,
+	service: string,
+	opts: LockOptions = {
+		force: false,
+		lockOverride: false,
+	},
+): Promise<void> {
+	const serviceLockFiles = lockFilesOnHost(appId, service);
+	for await (const file of serviceLockFiles) {
```

Comment:
Same as previous comment, the `for await` has no effect on promises in the loop here, and you're already using `await` for promises within the loop, so the `await isn't needed here.